Cogent Core demo crash

### Describe the bug

demo app crashes on linux

### How to reproduce

App crashes during testing Collections

### Example code

_No response_

### Relevant output

```shell
Crash log saved in /home/airbass/.config/Cogent\ Core\ Demo/crash-logs

Platform: Linux
System platform: Linux
App version: dev
Core version: dev
Time: 2024-07-31 11:56:53

panic: reflect: NumField of non-struct type string

goroutine 74 [running]:
runtime/debug.Stack()
	/usr/local/go/src/runtime/debug/stack.go:24 +0x5e
cogentcore.org/core/core.handleRecover({0xe0e760, 0xc007f1be20})
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/recover.go:38 +0x6c
cogentcore.org/core/core.(*renderWindow).eventLoop.func1()
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/renderwindow.go:393 +0x24
panic({0xe0e760?, 0xc007f1be20?})
	/usr/local/go/src/runtime/panic.go:770 +0x132
reflect.(*rtype).NumField(0xdf2040?)
	/usr/local/go/src/reflect/type.go:769 +0x59
cogentcore.org/core/base/reflectx.walkFields({0xe0e760?, 0xc001fa3780?, 0x711947?}, 0x0, 0xc00004f2e8, 0xc00004f2d0)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/base/reflectx/structs.go:36 +0xf3
cogentcore.org/core/base/reflectx.WalkFields(...)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/base/reflectx/structs.go:31
cogentcore.org/core/core.(*Form).getStructFields(0xc0029d0008)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/form.go:92 +0x115
cogentcore.org/core/core.(*Form).Init.func2(0xc002b46ba0)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/form.go:136 +0x5c
cogentcore.org/core/tree.(*NodeBase).Make.func1({0xc00050e740, 0x1, 0x0?})
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/tree/plan.go:103 +0x39
cogentcore.org/core/base/tiered.(*Tiered[...]).Do(0x416f05, 0xc00004f508?)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/base/tiered/tiered.go:30 +0x4c
cogentcore.org/core/tree.(*NodeBase).Make(0x3?, 0x2?)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/tree/plan.go:101 +0x5e
cogentcore.org/core/tree.(*NodeBase).UpdateFromMake(0xc0029d0008)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/tree/plan.go:111 +0x71
cogentcore.org/core/core.(*WidgetBase).UpdateWidget.(*NodeBase).RunUpdaters.func1({0xc0029cbbf0, 0xf5e4e0?, 0x0?})
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/tree/plan.go:121 +0x26
cogentcore.org/core/base/tiered.(*Tiered[...]).Do(0xf5e4e0, 0x1396928?)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/base/tiered/tiered.go:30 +0x4c
cogentcore.org/core/tree.(*NodeBase).RunUpdaters(...)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/tree/plan.go:119
cogentcore.org/core/core.(*WidgetBase).UpdateWidget(0xc0029d0008)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/update.go:25 +0x45
cogentcore.org/core/core.(*WidgetBase).Update.func1({0x724c1d687d68, 0xc0029d0008}, 0xc00004f6a8?)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/update.go:61 +0x25
cogentcore.org/core/core.(*WidgetBase).Update.(*WidgetBase).WidgetWalkDown.func2({0x724c1d687d18, 0xc0029d0008})
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/widget.go:449 +0x9c
cogentcore.org/core/tree.(*NodeBase).WalkDown(0xc0029c3508, 0xc007b007a0)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/tree/nodebase.go:500 +0x12b
cogentcore.org/core/core.(*WidgetBase).WidgetWalkDown(...)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/widget.go:447
cogentcore.org/core/core.(*WidgetBase).Update(0xc0029c3508)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/update.go:60 +0xad
cogentcore.org/core/core.(*Form).Init.func2.3.2({0x14de1d0?, 0xc002b40c40?})
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/form.go:240 +0x90
cogentcore.org/core/events.(*Listeners).Call(0xc00004f910, {0x14de1d0, 0xc002b40c40}, {0xc00004f8f8, 0x1, 0x66aa0a5b?})
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/events/listeners.go:61 +0xfd
cogentcore.org/core/core.(*WidgetBase).HandleEvent.func1(0xc00193d140)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/widgetevents.go:244 +0x65
cogentcore.org/core/base/tiered.(*Tiered[...]).Do(0x214cccd, 0xc00004f988?)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/base/tiered/tiered.go:30 +0x39
cogentcore.org/core/core.(*WidgetBase).HandleEvent(0xc0029cca88, {0x14de1d0?, 0xc002b40c40?})
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/widgetevents.go:243 +0x168
cogentcore.org/core/core.(*WidgetBase).Send(0xc0029cca88, 0x24, {0xc00004fa18?, 0xea?, 0x1?})
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/widgetevents.go:180 +0x77
cogentcore.org/core/core.(*WidgetBase).SendChange(...)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/widgetevents.go:187
cogentcore.org/core/core.(*Switch).Init.func2({0x14de1d0, 0xc002b40a80})
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/switch.go:133 +0xa6
cogentcore.org/core/events.(*Listeners).Call(0xc00004fac8, {0x14de1d0, 0xc002b40a80}, {0xc00004fab0, 0x1, 0x66aa0a5b?})
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/events/listeners.go:61 +0xfd
cogentcore.org/core/core.(*WidgetBase).HandleEvent.func1(0xc00193d1a0)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/widgetevents.go:244 +0x65
cogentcore.org/core/base/tiered.(*Tiered[...]).Do(0x214cccd, 0xc00004fb40?)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/base/tiered/tiered.go:31 +0x4c
cogentcore.org/core/core.(*WidgetBase).HandleEvent(0xc0029cca88, {0x14de1d0?, 0xc002b40a80?})
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/widgetevents.go:243 +0x168
cogentcore.org/core/core.(*WidgetBase).Send(0xc0029cca88, 0x5, {0xc00004fdf0?, 0x10002b3f8d0?, 0x724c140a5b58?})
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/widgetevents.go:180 +0x77
cogentcore.org/core/core.(*Events).handlePosEvent(0xc0001b6b00, {0x14de450, 0xc003921c00})
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/events.go:471 +0x209e
cogentcore.org/core/core.(*Events).handleEvent(0xc0001b6b00, {0x14de450, 0xc003921c00})
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/events.go:169 +0x58
cogentcore.org/core/core.(*Stage).mainHandleEvent(0xc0031314a0, {0x14de450, 0xc003921c00})
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/mainstage.go:372 +0x170
cogentcore.org/core/core.(*stages).mainHandleEvent(0xc003f92100, {0x14de450, 0xc003921c00})
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/mainstage.go:380 +0x6e
cogentcore.org/core/core.(*renderWindow).handleEvent(0xc003f920d0, {0x14de450, 0xc003921c00})
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/renderwindow.go:444 +0x15e
cogentcore.org/core/core.(*renderWindow).eventLoop(0xc003f920d0)
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/renderwindow.go:407 +0x87
created by cogentcore.org/core/core.(*renderWindow).goStartEventLoop in goroutine 1
	/home/airbass/go/pkg/mod/cogentcore.org/core@v0.3.1-0.20240731002359-25447b5a3c02/core/renderwindow.go:378 +0x65
```


### Platform

Linux

--- a/core/form.go
+++ b/core/form.go
@@ -157,7 +157,12 @@ func (fm *Form) Init() {
 			if !reflectx.Underlying(f.value).IsValid() {
 				typnm = "invalid"
 			}
-			valnm := fmt.Sprintf("value-%s-%s", f.path, typnm)
+			// We must have a different name for different indexes so that the index
+			// is always guaranteed to be accurate, which is required since we use it
+			// as the ground truth everywhere. The index could otherwise become invalid,
+			// such as when a ShouldDisplayer condition is newly satisfied
+			// (see https://github.com/cogentcore/core/issues/1096).
+			valnm := fmt.Sprintf("value-%s-%s-%d", f.path, typnm, i)
 			readOnlyTag := f.field.Tag.Get("edit") == "-"
 			def, hasDef := f.field.Tag.Lookup("default")
 
@@ -169,12 +174,16 @@ func (fm *Form) Init() {
 				w.Styler(func(s *styles.Style) {
 					s.SetTextWrap(false)
 				})
+				// TODO: technically we should recompute doc, readOnlyTag,
+				// def, hasDef, etc every time, as this is not fully robust
+				// (see https://github.com/cogentcore/core/issues/1098).
 				doc, _ := types.GetDoc(f.value, f.parent, f.field, label)
 				w.SetTooltip(doc)
 				if hasDef {
 					w.SetTooltip("(Default: " + def + ") " + w.Tooltip)
 					var isDef bool
 					w.Styler(func(s *styles.Style) {
+						f := fm.structFields[i]
 						isDef = reflectx.ValueIsDefault(f.value, def)
 						dcr := "(Double click to reset to default) "
 						if !isDef {
@@ -188,6 +197,7 @@ func (fm *Form) Init() {
 						}
 					})
 					w.OnDoubleClick(func(e events.Event) {
+						f := fm.structFields[i]
 						if isDef {
 							return
 						}
@@ -225,6 +235,7 @@ func (fm *Form) Init() {
 					wb.SetTooltip("(Default: " + def + ") " + wb.Tooltip)
 				}
 				wb.OnInput(func(e events.Event) {
+					f := fm.structFields[i]
 					fm.Send(events.Input, e)
 					if f.field.Tag.Get("immediate") == "+" {
 						wb.SendChange(e)
@@ -244,7 +255,8 @@ func (fm *Form) Init() {
 				wb.Updater(func() {
 					wb.SetReadOnly(fm.IsReadOnly() || readOnlyTag)
 					if i < len(fm.structFields) {
-						Bind(reflectx.UnderlyingPointer(fm.structFields[i].value).Interface(), w)
+						f := fm.structFields[i]
+						Bind(reflectx.UnderlyingPointer(f.value).Interface(), w)
 						vc := joinValueTitle(fm.ValueTitle, label)
 						if vc != wb.ValueTitle {
 							wb.ValueTitle = vc + " (" + wb.ValueTitle + ")"

